sql/sys_vars.cc

```diff
@@ -1869,6 +1869,21 @@ static bool check_charset(sys_var *, THD *thd, set_var *var) {
         my_error(ER_UNKNOWN_CHARACTER_SET, MYF(0), err.ptr());
         return true;
       }
+      // if 'default_collation_for_utf8mb4' is set to something other than
```

Comment:
What about similar code in check_collation_not_null() ?
In theory one can do `SET @@global.collation_connection = utf8mb4`.
Perhaps we should handle this case as well?
Or at least add comment why we don't think it is necessary/don't want to do it (indeed it might be non-trivial to distinguish cases when one uses charset name only -  `utf8mb4` vs case when one specifies collation explicitly -`utf8mb4_0900_ai_ci`.

Reply:
After a bit more pondering about it, I think it is a good idea to have test coverage for cases when user can use both charset name and collation name. Like `SET NAMES utf8mb4;` vs `SET NAMES utf8mb4 COLLATE utf8mb4_0900_ai_ci` and `SET @@global.collation_connection = utf8mb4` vs `SET @@global.collation_connection = utf8mb4_0900_ai_ci`.

Do you agree?

Reply:
@dlenev I extended coverage with the following statements
```sql
SET character_set_client = utf8mb4;
SET NAMES DEFAULT;
SET NAMES utf8mb4;
SET NAMES utf8mb4 COLLATE utf8mb4_general_ci;
SET CHARACTER SET DEFAULT;
SET CHARACTER SET utf8mb4;
SET collation_connection = utf8mb4_general_ci;
```
Note that 'SET NAMES utf8mb4 COLLATE utf8mb4_0900_ai_ci' will generate
`character_set_client=255` in the binary log - I believe this is correct behavior as uses pecifies the collation explicitly.

Likewise, 'SET collation_connection = utf8mb4_0900_ai_ci' will generate
`character_set_client=255` in the binary log.

As for 'SET collation_connection = utf8mb4' (when we try to assign a character set name to a collation variable), there is now problem here as this statatement is considered
syntactically incorrect In other words, no need to change `check_collation_not_null()`.